lib/Sema/TypeCheckType.cpp

```diff
@@ -2221,8 +2221,21 @@ Type TypeResolver::resolveAttributedType(TypeAttributes &attrs,
 
     for (auto i : FunctionAttrs) {
       if (attrs.has(i)) {
-        TC.diagnose(attrs.getLoc(i), diag::attribute_requires_function_type,
-                    TypeAttributes::getAttrName(i));
+          if (attrs.has(TAK_escaping)) {
+              auto &SM = TC.Context.SourceMgr;
+              auto loc = attrs.getLoc(TAK_escaping);
+              auto attrRange = SourceRange(
+                loc.getAdvancedLoc(-1),
+                Lexer::getLocForEndOfToken(SM, loc));
+              TC.diagnose(attrs.getLoc(i), diag::attribute_requires_function_type,
+                          TypeAttributes::getAttrName(i)).fixItRemove(attrRange);
+              if (ty->getOptionalObjectType()) {
+                  TC.diagnose(repr->getLoc(), diag::escaping_optional_type_argument);
+              }
+          } else {
+              TC.diagnose(attrs.getLoc(i), diag::attribute_requires_function_type,
```

Comment:
Can you unify this with the other case where we emit escaping_optional_type_argument, or extract the logic into a method? That code also has the getAdvancedLoc(-1)... business which could be replaced with getRangeWithAt().